server/etcdserver/server.go

```diff
@@ -1673,6 +1674,11 @@ func (s *EtcdServer) configure(ctx context.Context, cc raftpb.ConfChange) ([]*me
 			lg.Panic("failed to configure")
 		}
 		resp := x.(*confChangeResponse)
+		// etcdserver need to ensure the raft has already been notified
+		// or advanced before it responds to the client. Otherwise, the
+		// following config change request may be rejected.
+		// See https://github.com/etcd-io/etcd/issues/15528.
+		<-resp.raftAdvanceC
```

Comment:
This can work, but I'd suggest not to change the API layer. Let's get all the details encapsulated on the raft & etcd application layer, and the API layer should be agnostic to the raft & application side details. 

Please try to follow my original PR https://github.com/etcd-io/etcd/pull/15703/files

Reply:
The original PR waits for the added `raftAdvanceC`, however `raftAdvanceC` will not be sent because it is blocked until 'applyAll' calls 'applyWait.Trigger'. It's a deadlock and will be stuck forever. 

https://github.com/etcd-io/etcd/blob/6cdc9ae4fe1258247de1f7de372b3e8f0f0ea555/server/etcdserver/raft.go#L293-L310

It can be easily reproduced in [TestEnableAuth](https://github.com/etcd-io/etcd/blob/e9edc5432cffce26d53906528ecd125dfae3a699/server/embed/auth_test.go#L24) with https://github.com/etcd-io/etcd/commit/2d4c8eca3b3e51e2b3c43beefa35279955ea7c3d. In a single-node cluster, its first entry and committed entry type is `confChangeAdd` will be blocked in apply forever.

Reply:
I am open to considering other solutions, but I believe that it should be fixed just like this PR.

Reply:
Got it, thx for the clarification. 

But I think the implementation is slowly rotting. It's getting harder to understand and maintain. We should consider simplifying the implementation per my above comment https://github.com/etcd-io/etcd/pull/15708#discussion_r1243435646.

Reply:
Agreed. I feel it's inevitable because of the `r.Advance()` is invoked after `r.transport.Send(msgs)`. However, I doubt the reason of this ordering and pending for investigation. 

Changing the order might be risky so I lean towards to this fix (just for now).